pkg/tracing/tracer_otel.go

```diff
@@ -0,0 +1,171 @@
+	AttributePerEventCountLimit int `cfg:"attribute_per_event_count_limit" default:"128"`
+	AttributePerLinkCountLimit  int `cfg:"attribute_per_link_count_limit" default:"128"`
+}
+
+type otelTracer struct {
+	cfg.AppId
+	tracer trace.Tracer
+}
+
+func NewOtelTracer(ctx context.Context, config cfg.Config, logger log.Logger) (Tracer, error) {
+	appId := cfg.AppId{}
+	appId.PadFromConfig(config)
+
+	settings := &OtelSettings{}
+	config.UnmarshalKey("tracing.otel", settings)
+
+	otelExporterFactory := TraceExporters[settings.Exporter]
+
+	exporter, err := otelExporterFactory(ctx, config, logger)
+	if err != nil {
+		return nil, err
+	}
+
+	tracerProvider := sdktrace.NewTracerProvider(
+		sdktrace.WithBatcher(exporter),
+		sdktrace.WithResource(resource.NewWithAttributes(
+			semconv.SchemaURL,
+			semconv.ServiceName(fmt.Sprintf("%v-%v-%v-%v", appId.Project, appId.Environment, appId.Family, appId.Application)),
+		)),
+		sdktrace.WithSampler(sdktrace.ParentBased(sdktrace.TraceIDRatioBased(settings.SamplingRatio))),
+		sdktrace.WithRawSpanLimits(sdktrace.SpanLimits{
+			AttributeValueLengthLimit:   settings.AttributeValueLengthLimit,
+			AttributeCountLimit:         settings.AttributeCountLimit,
+			EventCountLimit:             settings.EventCountLimit,
+			LinkCountLimit:              settings.LinkCountLimit,
+			AttributePerEventCountLimit: settings.AttributePerEventCountLimit,
+			AttributePerLinkCountLimit:  settings.AttributePerLinkCountLimit,
+		}),
+	)
+
+	otel.SetTracerProvider(tracerProvider)
+	otel.SetTextMapPropagator(propagation.TraceContext{})
+
+	tracer := otel.GetTracerProvider().Tracer(
+		instrumentationName,
+		trace.WithInstrumentationVersion(instrumentationVersion),
+		trace.WithSchemaURL(semconv.SchemaURL),
+	)
+
+	return NewOtelTracerWithInterfaces(appId, tracer), nil
+}
+
+func NewOtelTracerWithInterfaces(appId cfg.AppId, tracer trace.Tracer) *otelTracer {
+	return &otelTracer{
+		AppId:  appId,
+		tracer: tracer,
+	}
+}
+
+func (t *otelTracer) StartSubSpan(ctx context.Context, name string) (context.Context, Span) {
+	ctx, span := t.tracer.Start(ctx, name)
+
+	return newOtelSpan(ctx, span)
+}
+
+func (t *otelTracer) StartSpan(name string) (context.Context, Span) {
+	return t.StartSubSpan(context.Background(), name)
+}
+
+func (t *otelTracer) StartSpanFromContext(ctx context.Context, name string) (context.Context, Span) {
+	if parentSpan := GetSpanFromContext(ctx); parentSpan != nil {
+		parentTrace := parentSpan.GetTrace()
+		return t.spanFromTrace(ctx, parentTrace, name)
+	}
+
+	if ctxTrace := GetTraceFromContext(ctx); ctxTrace != nil {
+		return t.spanFromTrace(ctx, ctxTrace, name)
+	}
+
+	return t.StartSubSpan(ctx, name)
+}
+
+func (t *otelTracer) HttpHandler(h http.Handler) http.Handler {
+	name := fmt.Sprintf("%v-%v-%v-%v", t.Project, t.Environment, t.Family, t.Application)
+	handlerFunc := http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		ctx := r.Context()
+		span := trace.SpanFromContext(r.Context())
```

Comment:
TBH I do not think this is necessary. Maybe can be removed. If it is meaningful can you explain it here? This needs to be ended in the code as well I guess, otherwise I would go with also `defer`ing the `span.End()`

Reply:
* the span didn't start here and it will be ended in `go.opentelemetry.io/contrib/instrumentation/net/http/otelhttp@v0.44.0/handler.go:164`

* I checked the implementation and it seems that we need to store the span in the context for the case that someone uses `StartSpanFromContext` func to start a span